// configcat-common/lib/esm/ProjectConfig.js
var ProjectConfig = /** @class */ (function () {
    function ProjectConfig(timeStamp, jsonConfig, httpETag) {
        this.Timestamp = timeStamp;
        this.ConfigJSON = JSON.parse(jsonConfig);
        this.HttpETag = httpETag;
    }
    /**
     * Determines whether the specified ProjectConfig instances are considered equal.
     */
    ProjectConfig.equals = function (projectConfig1, projectConfig2) {
        if (!projectConfig1 || !projectConfig2)
            return false;
        return (this.ensureStrictEtag(projectConfig1.HttpETag) === this.ensureStrictEtag(projectConfig2.HttpETag));
    };
    ProjectConfig.ensureStrictEtag = function (etag) {
        if (!etag) {
            return etag;
        }
        if (etag.length > 2 && etag.substr(0, 2).toLocaleUpperCase() === "W/") {
            return etag.substring(2);
        }
        return etag;
    };
    return ProjectConfig;
}());
export { ProjectConfig };
var ConfigFile = /** @class */ (function () {
    function ConfigFile() {
    }
    ConfigFile.Preferences = "p";
    ConfigFile.FeatureFlags = "f";
    return ConfigFile;
}());
export { ConfigFile };
var Preferences = /** @class */ (function () {
    function Preferences() {
    }
    Preferences.BaseUrl = "u";
    Preferences.Redirect = "r";
    return Preferences;
}());
export { Preferences };
var Setting = /** @class */ (function () {
    function Setting() {
    }
    Setting.Value = "v";
    Setting.SettingType = "t";
    Setting.RolloutPercentageItems = "p";
    Setting.RolloutRules = "r";
    Setting.VariationId = "i";
    return Setting;
}());
export { Setting };
var RolloutRules = /** @class */ (function () {
    function RolloutRules() {
    }
    RolloutRules.Order = "o";
    RolloutRules.ComparisonAttribute = "a";
    RolloutRules.Comparator = "t";
    RolloutRules.ComparisonValue = "c";
    RolloutRules.Value = "v";
    RolloutRules.VariationId = "i";
    return RolloutRules;
}());
export { RolloutRules };
var RolloutPercentageItems = /** @class */ (function () {
    function RolloutPercentageItems() {
    }
    RolloutPercentageItems.Order = "o";
    RolloutPercentageItems.Value = "v";
    RolloutPercentageItems.Percentage = "p";
    RolloutPercentageItems.VariationId = "i";
    return RolloutPercentageItems;
}());
export { RolloutPercentageItems };
